src/Data/Singletons/Deriving/Ord.hs

```diff
@@ -24,7 +24,11 @@ import Data.Singletons.Syntax
 mkOrdInstance :: Quasi q => DType -> [DCon] -> q UInstDecl
 mkOrdInstance ty cons = do
   let constraints = inferConstraints (DConPr ordName) cons
-  compare_eq_clauses <- mapM mk_equal_clause cons
+  -- This case also handles empty datatypes, since we can just create
+  -- one clause with wildcard matches and zero fields to fold over
+  compare_eq_clauses <- mapM mk_equal_clause $ if null cons
```

Comment:
The use of `Maybe` here seems a bit off. Instead, I would keep this (and the next line, for `compare_noneq_clauses`) as they are, and instead just generate a different `UFunction` at the last step. Seems simpler.